Refactor TimeSeries.q_transform() to support eventgrams

alurban opened the pull request:
This PR substantially refactors `TimeSeries.q_transform()` to support eventgrams, a feature ported over from [**`gwdetchar-omega`**](https://github.com/gwdetchar/gwdetchar/blob/master/bin/gwdetchar-omega#L166-L254). To simplify review, here is a summary of changes made:

**New Features**
* Introduce a new object class, `~gwpy.signal.qtransform.QGram`, to support eventgrams
* Introduce a new function, `~gwpy.signal.qtransform.q_scan()`, which identifies the time-frequency plane at fixed Q with the most significant tile within a given (optional) search window. This method returns an instance of `QGram` along with the estimated false alarm rate from white Gaussian noise.
* Include a method, `QGram.table()`, which converts a populated `QGram` object to an `EventTable` so that it can be rendered with matplotlib. Table columns are `'time'`, `'duration'`, `'frequency'`, `'bandwidth'`, and `'energy'`. The only argument is an SNR threshold, which selects for tiles with energy greater than or equal `snrthresh**2/2`.
* Introduce a `TimeSeries.q_gram` method to wrap around `QGram.table`.
* Add a `mismatch` argument for `q_scan()` and `TimeSeries.q_transform()`.
* Add a unit test for the `QGram` object under `gwpy/signal/tests/test_qtransform.py`.
* Add an example for the docs under `examples/signal/qscan.py`.

**Refactor Existing Features**
* Refactor spectrogram interpolation into `~gwpy.signal.qtransform.QGram.interpolate()`. This reduces the cyclomatic complexity of `TimeSeries.q_transform()` to a level acceptable to CodeClimate.

This fixes #922 and fixes #995.

cc @duncanmmacleod, @areeda, @scottcoughlin2014

pep8speaks commented:
Hello @alurban! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`gwpy/cli/tests/test_qtransform.py`](https://github.com/alurban/gwpy/blob/0fbee8cc505d6279da613ec79869392196d2b78f/gwpy/cli/tests/test_qtransform.py) !

 - There are no PEP8 issues in the file [`gwpy/signal/qtransform.py`](https://github.com/alurban/gwpy/blob/0fbee8cc505d6279da613ec79869392196d2b78f/gwpy/signal/qtransform.py) !

 - There are no PEP8 issues in the file [`gwpy/timeseries/tests/test_timeseries.py`](https://github.com/alurban/gwpy/blob/0fbee8cc505d6279da613ec79869392196d2b78f/gwpy/timeseries/tests/test_timeseries.py) !

 - There are no PEP8 issues in the file [`gwpy/timeseries/timeseries.py`](https://github.com/alurban/gwpy/blob/0fbee8cc505d6279da613ec79869392196d2b78f/gwpy/timeseries/timeseries.py) !

alurban commented:
@duncanmmacleod, since #983 is merged, I've resolved WIP status here. This PR is now ready for review.

duncanmmacleod commented:
@alurban, please rebase against `gwpy/gwpy@develop` to correct the appveyor failures.

alurban commented:
@duncanmmacleod, this PR has been updated to pass build tests.

duncanmmacleod commented:
@alurban, unfortunately, I'm not happy with this change - I can't 100% put my finger on why.

I do like the modularisation of the `TimeSeries.q_transform` method into the `gwpy.signal.qtransform` module, but i don't like the overloading of the `QPlane` object as a container for the _output_ of a Q-transform scan operation.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1749,11 +1749,122 @@ def notch(self, frequency, type='iir', filtfilt=True, **kwargs):
                                   type=type, **kwargs)
         return self.filter(*zpk, filtfilt=filtfilt)
 
+    def q_gram(self, qrange=(4, 64), frange=(0, numpy.inf), gps=None,
```
What purpose does this method serve in the `TimeSeries` object? I think it might be better exposed simply via the `gwpy.signal.qtransform` module for expert users (like yourself) to use since the returned object doesn't really have much utility, except as a stepping-stone towards an `EventTable` or a `Spectrogram`.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1805,13 +1920,17 @@ def q_transform(self, qrange=(4, 64), frange=(0, numpy.inf),
+        qgram : `~gwpy.signal.qtransform.QPlane`, optional
+            ready-made time-frequency plane (at fixed Q) over which to
+            interpolate, will override `whiten` if passed, default: `None`
+
```
This keyword really doesn't belong here, since its inclusion removes all usage of the parent `TimeSeries` object in the method and reduces it to an alias of `qgram.interpolate`. If the user already has a `<qgram>` object, they can call the `interpolate` method themselves?

alurban commented:
@duncanmmacleod, I've implemented the idea you suggested to me on Slack. Under these new changes, there's now an object class called `QGram` and a utility called `~gwpy.signal.qtransform.q_scan()` that support eventgrams, which are wrapped around by `TimeSeries.q_transform()`. I've also updated gwdetchar/gwdetchar#206 to reflect this.

duncanmmacleod commented on gwpy/signal/qtransform.py:
```diff
@@ -149,6 +157,46 @@ def __iter__(self):
+            significant tile
+
+        N : `int`
+            estimated number of statistically significant tiles
```
What's the utility of this number? Can this be computed externally (i.e. not included here)?

alurban replied:
@duncanmmacleod, this number is needed to compute the false alarm rate (FAR) expected from white Gaussian noise, which is done in the [**`q_scan` utility**](https://github.com/alurban/gwpy/blob/eventgram/gwpy/signal/qtransform.py#L660). At least one downstream consumer, `gwdetchar-omega`, uses this metric to filter out quiet or insignificant channels, and other consumers (e.g., commissioners writing data quality checks) may also find it useful.

The number of statistically independent tiles is really a property of the search over a `QTiling`, i.e., it's calculated as a sum over all time-frequency planes in the tiling used in a search. So it needs to be made available as part of the `QTiling` object somehow, and this seemed to me like a natural way to do that.

alurban commented:
thanks @duncanmmacleod, I can definitely do that, do you want it in the docstring for `q_scan` or as a separate tutorial under `examples/`?

duncanmmacleod commented:
@alurban, I was meaning under `examples/`.

alurban commented:
@duncanmmacleod, I've added an example under `examples/signal/qscan.py` that computes a raw Q-scan around GW170817.

alurban commented:
@duncanmmacleod, I've updated this PR (and the example plot) to incorporate a new method, `TimeSeries.q_gram()`, which wraps around `QGram.table()`. This doesn't handle auto-whitening, but I personally am 100% fine with that, not least because the worked example warns the user to whiten beforehand.

duncanmmacleod commented on gwpy/signal/qtransform.py:
```diff
@@ -377,7 +435,227 @@ def transform(self, fseries, norm=True, epoch=None):
+        names = ('time', 'frequency', 'duration', 'bandwidth', 'energy')
+        rec = numpy.recarray((0,), names=names, formats=['f8'] * len(names))
+        for f, bw, row in zip(freqs, bws, self.energies):
+            ind, = (row.value >= snrthresh ** 2 / 2).nonzero()
```
@alurban, I'm concerned here that if `snrthresh` is given as an `int`, then `snrthresh ** 2 / 2` is entirely composed of integers and might not do what you think it does for python < 3: e.g.

```python
>>> 5 ** 2 / 2
12
>>> 5 ** 2 / 2.
12.5
```

To be safe, can you please add a float decimal point:

```suggestion
            ind, = (row.value >= snrthresh ** 2. / 2.).nonzero()
```
